extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java

```diff
@@ -158,6 +166,7 @@ protected synchronized TaskStatus join(long timeout) throws IllegalStateExceptio
           state.get(),
           State.RUNNING
       );
+      stateListener.accept(state.get(), taskId.getOriginalTaskId());
```

Comment:
Instead of calling this everywhere, you could add a new method `updateState` which does a `state.compareAndSet` and invokes the `stateListener`.

Reply:
Updated.